### PR TITLE
Karla G - Sockets

### PR DESCRIPTION
**Pull request by kguadron**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | An initial design challenge I faced was deciding which actions (methods) belong to which classes. I encountered these same questions in wave three when creating the hotel block.
What was a design decision you made that changed over time over the project? | When I planned out wave three, I intended to have a BlockReservation class that inherited from the Reservation class. I changed my mind as I built the Block class.
What was a concept you gained clarity on, or a learning that you'd like to share? | The whole process of designing and building a system is much clearer to me. At the beginning of the project, I felt lost with that and by Wave 3 I was able to make decisions with more intentionality.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | An example of a nominal test I wrote was testing to see if objects were an instance of the intended class.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | An edge case I tested for my reserve_block_room was checking if the block id existed or not.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I further developed that skill during this project. Following the trend I explained before, during the beginning phases it felt more tenuous and by Wave 3 I pseudocoded everything and felt it really helped with executing the writing of the code.

### CODE REVIEW

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    def generate_rooms(number_of_rooms)
+      number_of_rooms.times do |i|
+        @all_rooms << Room.new(number: i + 1)
+      end
+    end
+
+    def see_reservations_by_date(date)
+      this_dates_reservations = []
+      @all_reservations.each do |reservation|
+        this_dates_reservations << reservation if reservation.start_date == date
```
Comment (droberts-sea) [inline lib/hotel.rb:26]: The `.select` enumerable might be helpful to clean up this code.

File: lib/hotel.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      end
+      return this_dates_reservations
+    end
+
+    def see_available_rooms_by_date(possible_start_date, possible_end_date)
+      these_dates_available_rooms = []
+      @all_rooms.each do |room|
+        these_dates_available_rooms << room if room.date_available?(possible_start_date, possible_end_date)
```
Comment (droberts-sea) [inline lib/hotel.rb:34]: On line 34, I like that you call `room.date_available?` instead of having a bunch of date math here in `Hotel`. This is what Metz is talking about when she says you should "ask for _what_ instead of telling _how_".

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        reservation = HotelSystem::Reservation.new(room: available_room, start_date: start_date, end_date: end_date, guest: guest)
+        available_room.reservations << reservation
+        @all_reservations << reservation
+      end
+    end
+
+    def room_available?(new_start_date, new_end_date)
+      @all_rooms.each do |room|
+        return room if room.date_available?(new_start_date, new_end_date)
+      end
```
Comment (droberts-sea) [inline lib/hotel.rb:58]: In Ruby, ending a method name with a question mark implies that it returns either `true` or `false`, but this method may return an instance of `Room`. A better name might be `find_available_room`.

File: lib/hotel.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      return false
+    end
+
+    def hold_block(start_date, end_date, room_collection)
+      room_collection.each do |room|
+        raise ArgumentError, "Room #{room.number} is already in a block" if room.block_id != nil
+      end
```
Comment (droberts-sea) [inline lib/hotel.rb:65]: I don't know that I agree with the idea of storing the `block_id` as an instance variable on the `Room`. That means that each room can only ever be included in one block. However, blocks don't last for ever - they're for a set amount of time, and on other dates the room should be free to be included in other blocks.

File: lib/hotel.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      @all_blocks << block
+    end
+
+    def reserve_block_room(start_date, end_date, room, block_id, guest)
+      desired_block = @all_blocks.detect { |block| block.id == block_id }
+
+      if desired_block == nil
+        raise ArgumentError, "The block entered does not exist or it does not have any rooms available for reseervation"
+      elsif desired_block.room_still_available?
+        reservation = HotelSystem::Reservation.new(room: room, start_date: start_date, end_date: end_date, guest: guest)
```
Comment (droberts-sea) [inline lib/hotel.rb:84]: This will not raise an exception if the block is full. To do so, you would need code like this:

```ruby
if desired_block == nil || desired_block.room_still_available?
  raise ArgumentError #...
else
  # ...
```

However, I would prefer you split the two cases and provide a little more detailed information in your error message.

File: lib/reservation.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+      @price = 200
+      @cost = calculate_cost
+      @guest = guest
+      @reservation_id = ((0...5).map { rand(10) }).join.to_s
+      raise ArgumentError, "Invalid date range" if @end_date <= @start_date
+    end
+
+    def calculate_cost
+      return ("%.2f" % ((@end_date - @start_date) * @price)).to_f
+    end
```
Comment (droberts-sea) [inline lib/reservation.rb:21]: I like the decision to make this a method rather than storing the cost as an instance variable. That way if the number of nights or the cost per night were to change (say in a future version of the program), the total cost will automatically be correct.

File: lib/room.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+      @reservations = []
+      @block_id = block_id
+    end
+
+    def date_available?(possible_start_date, possible_end_date)
+      @reservations.each do |reservation|
+        if possible_start_date == reservation.start_date
+          return false
+        elsif reservation.start_date < possible_start_date && possible_start_date < reservation.end_date
+          return false
```
Comment (droberts-sea) [inline lib/room.rb:18]: You can simplify all of this complex conditional logic into one line! See the [instructor implementation](https://github.com/droberts-ada/hotel/blob/dpr/solution/lib/date_range.rb#L17).

Each of these would make an excellent test case though.

File: spec/hotel_spec.rb [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+      )
+      expect(room.reservations.length).must_equal 1
+      expect(room.reservations[0]).must_be_kind_of HotelSystem::Reservation
+    end
+
+    it "can assign a room, that doesn't have other conflicting reservations, to the reservation if the selected room is not available" do
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 10),
+        Date.new(2019, 3, 13),
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:140]: You do a lot of repeated setup work throughout these tests. Would it be possible to consolidate this into a shared `before` block or something similar?

File: spec/hotel_spec.rb [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+          "Zoe"
+        )
+      end.must_raise ArgumentError
+    end
+  end
+
+  describe "room_available? method" do
+    before do
+      @hotel = HotelSystem::Hotel.new
+      @hotel.all_rooms.slice!(2..19)
```
Comment (droberts-sea) [inline spec/hotel_spec.rb:186]: This is a clever way to make this test work. Another option would be to pass in the number of rooms to the `Hotel` constructor, possibly with a default value of 20.

File: spec/room_spec.rb [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        end_date: Date.new(2019, 3, 14),
+        guest: "Sam",
+      )
+      @room.reservations << @reservation
+    end
+
+    it "returns false if the possible start date matches the start date of the reservation" do
+      expect(@room.date_available?(Date.new(2019, 3, 11), Date.new(2019, 3, 12))).must_equal false
+      expect(@room.date_available?(Date.new(2019, 3, 11), Date.new(2019, 3, 14))).must_equal false
+      expect(@room.date_available?(Date.new(2019, 3, 11), Date.new(2019, 3, 15))).must_equal false
```
Comment (droberts-sea) [inline spec/room_spec.rb:31]: Good work nailing all of these test cases.

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | yes
Answer comprehension questions | yes
**Design** |
Each class is responsible for a single piece of the program | yes
Classes are loosely coupled | yes
**Wave 1** |
List rooms | yes
Reserve a room fr a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | yes
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | yes
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | yes
**Fundamentals** |
Names variables, classes and modules appropriately | mostly
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and `Enumerable` methods | mostly
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a namespace | yes
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | Excellent job overall. This code is well-organized, well-written, and well-tested. I've left a number of inline comments, but in general I like what I see. Keep up the hard work!
